hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java

```diff
@@ -0,0 +1,171 @@
+ */
+final class SlowOperationDetector extends Thread {
+
+    private static final long ONE_SECOND_IN_NANOS = TimeUnit.SECONDS.toNanos(1);
+
+    private final BasicOperationScheduler.OperationThread[] genericOperationThreads;
+    private final BasicOperationScheduler.OperationThread[] partitionOperationThreads;
+
+    private final CurrentOperationData[] genericCurrentOperationData;
+    private final CurrentOperationData[] partitionCurrentOperationData;
+
+    private final long slowOperationThresholdNanos;
+    private final long logPurgeIntervalNanos;
+    private final long logLifetimeNanos;
+
+    private final ILogger logger;
+
+    private final ConcurrentHashMap<Integer, SlowOperationLog> slowOperationLogs = new ConcurrentHashMap<Integer, SlowOperationLog>();
+
+    private long nowNanos;
+    private long nowMillis;
+
+    private volatile boolean running = true;
+
+    public SlowOperationDetector(Node node, BasicOperationScheduler scheduler) {
+        super(SlowOperationDetector.class.getSimpleName());
+        setDaemon(true);
+
+        genericOperationThreads = scheduler.genericOperationThreads;
+        partitionOperationThreads = scheduler.partitionOperationThreads;
+
+        genericCurrentOperationData = new CurrentOperationData[genericOperationThreads.length];
+        partitionCurrentOperationData = new CurrentOperationData[partitionOperationThreads.length];
+
+        initCurrentOperationData(genericCurrentOperationData);
+        initCurrentOperationData(partitionCurrentOperationData);
+
+        int thresholdMillis = Math.abs(node.groupProperties.SLOW_OPERATION_DETECTOR_THRESHOLD_MILLIS.getInteger());
+        int purgeIntervalSeconds = Math.abs(node.groupProperties.SLOW_OPERATION_DETECTOR_LOG_PURGE_INTERVAL_SECONDS.getInteger());
+        int logLifetimeSeconds = Math.abs(node.groupProperties.SLOW_OPERATION_DETECTOR_LOG_RETENTION_SECONDS.getInteger());
+
+        slowOperationThresholdNanos = TimeUnit.MILLISECONDS.toNanos(thresholdMillis);
+        logPurgeIntervalNanos = TimeUnit.SECONDS.toNanos(purgeIntervalSeconds);
+        logLifetimeNanos = TimeUnit.SECONDS.toNanos(logLifetimeSeconds);
+
+        logger = node.getLogger(BasicOperationScheduler.class);
+    }
+
+    @Override
+    public void run() {
+        long lastLogPurge = System.nanoTime();
+
+        while (running) {
+            nowNanos = System.nanoTime();
+            nowMillis = System.currentTimeMillis();
+
+            checkForSlowOperations(genericOperationThreads, genericCurrentOperationData);
+            checkForSlowOperations(partitionOperationThreads, partitionCurrentOperationData);
+
+            if (nowNanos - lastLogPurge > logPurgeIntervalNanos) {
+                lastLogPurge = nowNanos;
+                purgeLogs();
+            }
+
+            try {
+                TimeUnit.NANOSECONDS.sleep(ONE_SECOND_IN_NANOS - (System.nanoTime() - nowNanos));
+            } catch (InterruptedException ignored) {
+                break;
```

Comment:
You don't need to break. The wile loop will decide if the thread needs to stop. So I would just remove the break;